src/solver/exprtype.go

```diff
@@ -14,13 +14,23 @@ import (
 	"github.com/z7zmey/php-parser/node/scalar"
 )
 
+// binaryMathOpType is used for binary arithmetic operations
 func binaryMathOpType(sc *meta.Scope, cs *meta.ClassParseState, left, right node.Node, custom []CustomType) *meta.TypesMap {
 	if ExprTypeLocalCustom(sc, cs, left, custom).IsInt() && ExprTypeLocalCustom(sc, cs, right, custom).IsInt() {
 		return meta.NewTypesMap("int")
 	}
 	return meta.NewTypesMap("float")
 }
 
+// binaryPlusOpType is a special case as "plus" is also used for array union operation
+func binaryPlusOpType(sc *meta.Scope, cs *meta.ClassParseState, left, right node.Node, custom []CustomType) *meta.TypesMap {
+	// TODO: PHP will raise fatal error if one operand is array and other is not, so we may check it too
+	if ExprTypeLocalCustom(sc, cs, left, custom).IsArray() && ExprTypeLocalCustom(sc, cs, right, custom).IsArray() {
+		return meta.NewTypesMap("mixed[]")
```

Comment:
You might want to return a union of concrete types here instead so that information about types isn't lost

Reply:
I would still suggest that you return the _union_ of the types (there is a MergeTypesMap method that does that) instead of using mixed.

Imagine the following (correct) PHP code:

<?php
class A { public function doSomething() { } }
class B { public function doSomething() { } }

$arrA = [ 'a' => new A(); ];
$artB = [ 'b' => new B(); ];
$arrUnion = $arrA + $arrB;

$arrA['a']->doSomething(); // no error
$arrB['b']->doSomething(); // no error
$arrUnion['a']->doSomething(); // this will be "no such method" mixed->doSomething () if you didn't merge the types

Reply:
MergeTypesMap returns "int[]|string[]" for [1] + ["foo"], which is wrong. Correct type for this union should be (int|string)[].

Not sure if code supports "(A|B)[]" types yet.

Reply:
You're right but this is how noverify works right now. It currently does treat a[]|b[] as (a|b)[] anyway